sql/expression/function/timediff.go

```diff
@@ -206,3 +208,168 @@ func (d *DateDiff) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
 func (d *DateDiff) String() string {
 	return fmt.Sprintf("DATEDIFF(%s, %s)", d.Left, d.Right)
 }
+
+// TimestampDiff returns expr1 − expr2 expressed as a value in unit specified.
+type TimestampDiff struct {
+	unit  sql.Expression
+	expr1 sql.Expression
+	expr2 sql.Expression
+}
+
+var _ sql.FunctionExpression = (*TimestampDiff)(nil)
+
+// NewTimestampDiff creates a new TIMESTAMPDIFF() function.
+func NewTimestampDiff(u, e1, e2 sql.Expression) sql.Expression {
+	return &TimestampDiff{u, e1, e2}
+}
+
+// FunctionName implements sql.FunctionExpression
+func (t *TimestampDiff) FunctionName() string {
+	return "timestampdiff"
+}
+
+// Description implements sql.FunctionExpression
+func (t *TimestampDiff) Description() string {
+	return "gets difference between two dates in result of units specified."
+}
+
+// Children implements the sql.Expression interface.
+func (t *TimestampDiff) Children() []sql.Expression {
+	return []sql.Expression{t.unit, t.expr1, t.expr2}
+}
+
+// Resolved implements the sql.Expression interface.
+func (t *TimestampDiff) Resolved() bool {
+	return t.unit.Resolved() && t.expr1.Resolved() && t.expr2.Resolved()
+}
+
+// IsNullable implements the sql.Expression interface.
+func (t *TimestampDiff) IsNullable() bool {
+	return t.unit.IsNullable() && t.expr1.IsNullable() && t.expr2.IsNullable()
+}
+
+// Type implements the sql.Expression interface.
+func (t *TimestampDiff) Type() sql.Type { return sql.Int64 }
+
+// WithChildren implements the Expression interface.
+func (t *TimestampDiff) WithChildren(children ...sql.Expression) (sql.Expression, error) {
+	if len(children) != 3 {
+		return nil, sql.ErrInvalidChildrenNumber.New(t, len(children), 3)
+	}
+	return NewTimestampDiff(children[0], children[1], children[2]), nil
+}
+
+// Eval implements the sql.Expression interface.
+func (t *TimestampDiff) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
+	expr1, err := t.expr1.Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+	if expr1 == nil {
+		return nil, nil
+	}
+
+	expr2, err := t.expr2.Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+	if expr2 == nil {
+		return nil, nil
+	}
+
+	expr1, err = sql.Datetime.Convert(expr1)
+	if err != nil {
+		return nil, err
+	}
+
+	expr2, err = sql.Datetime.Convert(expr2)
+	if err != nil {
+		return nil, err
+	}
+
+	unit, err := t.unit.Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+	if unit == nil {
+		return nil, nil
+	}
+
+	unit = strings.TrimPrefix(strings.ToLower(unit.(string)), "sql_tsi_")
```

Comment:
What if this does not contain the prefix sql_tsi?

Reply:
According to mysql docs, "The unit value may be specified using one of keywords as shown, or with a prefix of SQL_TSI_. For example, DAY and SQL_TSI_DAY both are legal." So, it just handles both `DAY` and `SQL_TSI_DAY` the same way. It removes `SQL_TSI_` if it's there, otherwise, it won't affect the unit variable

Reply:
Ah my mistake. For some reason I thought you didn't handle any adversarial inputs. Please add a test case for invalid unit values

Reply:
(if it already doesn't exist)

Reply:
got it!